src/main/java/org/truffleruby/core/string/StringNodes.java

```diff
@@ -754,106 +759,252 @@ public static boolean bothSingleByteOptimizable(DynamicObject string, DynamicObj
                 @Cached("create()") BranchProfile errorProfile) {
-            if (args.length == 0) {
+            if (ropes.length == 0) {
                 errorProfile.enter();
                 throw new RaiseException(coreExceptions().argumentErrorEmptyVarargs(this));
             }
 
-            DynamicObject[] otherStrings = new DynamicObject[args.length];
-
-            for (int i = 0; i < args.length; i++) {
-                otherStrings[i] = toStr.executeToStr(frame, args[i]);
-            }
-
-            return countSlow(string, otherStrings);
+            Encoding enc = findEncoding(string, ropes);
+            return countSlow(string, ropes, enc);
         }
 
         @TruffleBoundary
-        private int countSlow(DynamicObject string, DynamicObject... otherStrings) {
+        private int countSlow(DynamicObject string, Rope[] ropes, Encoding enc) {
             assert RubyGuards.isRubyString(string);
 
-            DynamicObject otherStr = otherStrings[0];
-            Encoding enc = checkEncodingNode.executeCheckEncoding(string, otherStr);
-
             final boolean[]table = new boolean[StringSupport.TRANS_SIZE + 1];
-            StringSupport.TrTables tables = StringSupport.trSetupTable(rope(otherStr), table, null, true, enc);
-            for (int i = 1; i < otherStrings.length; i++) {
-                otherStr = otherStrings[i];
+            StringSupport.TrTables tables = makeTables(string, ropes, table, enc);
+            return processStr(string, table, enc, tables);
+        }
+    }
 
-                assert RubyGuards.isRubyString(otherStr);
+    public abstract static class TrTableNode extends CoreMethodArrayArgumentsNode {
+        @Child protected ToStrNode toStr = ToStrNodeGen.create(null);
+        @Child protected EncodingNodes.NegotiateCompatibleRopeEncodingNode checkEncodingNode = EncodingNodes.NegotiateCompatibleRopeEncodingNode.create();
 
-                enc = checkEncodingNode.executeCheckEncoding(string, otherStr);
-                tables = StringSupport.trSetupTable(rope(otherStr), table, tables, false, enc);
+        protected boolean[] makeSqueeze() {
+            return new boolean[StringSupport.TRANS_SIZE + 1];
+        }
+
+        protected Encoding findEncoding(DynamicObject string, Rope[] ropes) {
+            Rope rope = StringOperations.rope(string);
+            Encoding enc = checkEncodingNode.executeNegotiate(rope, ropes[0]);
+            for (int i = 1; i < ropes.length; i++) {
+                enc = checkEncodingNode.executeNegotiate(rope, ropes[i]);
+            }
+            return enc;
+        }
+
+        protected TrTables makeTables(DynamicObject string, Rope[] ropes, boolean[] squeeze, Encoding enc) {
+            StringSupport.TrTables tables = StringSupport.trSetupTable(ropes[0], squeeze, null, true, enc);
+
+            for (int i = 1; i < ropes.length; i++) {
+                tables = StringSupport.trSetupTable(ropes[i], squeeze, tables, false, enc);
```

Comment:
Does this maintain the same semantics of the code you replaced? I haven't looked into the `trSetupTable` logic that much, but previously each new table transition also took the newly negotiated encoding into account. This new logic now gets the final negotiated encoding and uses that final value for each table transition.

Reply:
I believe this maintains the semantics, most cases where the encoding would change never actually occur because the encodings are incompatible.

Reply:
Would be worth adding a few spec to make sure this is still the case.

Reply:
I've not managed to add a useful new spec for this, (we'd really want a test down at the Java level of what `TrTables` were produced for different string and encoding combinations. But I have worked out the logic of how the ropes are being consumed, and hence why the change is okay, so I'm adding a comment explaining that.